lib/datetime/datetime.ex

```diff
@@ -39,7 +39,8 @@ defimpl Timex.Protocol, for: DateTime do
   def to_date(date), do: DateTime.to_date(date)
 
   @spec to_datetime(DateTime.t, timezone :: Types.valid_timezone) :: DateTime.t | {:error, term}
-  def to_datetime(%DateTime{} = d, _timezone), do: d
+  def to_datetime(%DateTime{time_zone: time_zone} = d, timezone) when time_zone == timezone, do: d
```

Comment:
You can just do `to_datetime(%DateTime{time_zone: timezone} = d, timezone), do: d` which will only match if both the provided timezone and the one on the DateTime are the same.